### DISCUSSION TITLE
Don't prebuild module at compile time (For reading env once, at runtime)

### DISCUSSION THREAD

**Discussion by bbergeron-xlkey** [vercel/next.js#70644] [Help] [answered]
### Summary

My authentication module reads from env at it's topmost level and throw an exception if certain variables are undefined. This works with other architectures, but fails with Next.js since the module is evaluated at build time, when these variables aren't defined yet. My first idea is to somehow tell Next.js to not prebuild this module (it doesn't contains any JSX anyway) so I can keep the same logic as I would have within our previous architecture. How can I do that? Any other suggestions?

### Example
```ts
// authentication.ts
import { NextRequest } from "next/server";
import jwt from "jsonwebtoken";

const jwtSecret = process.env.JWT_SECRET;
if (jwtSecret === undefined)
    // Should only crash the app at runtime, but is currently crashing at build time
    throw new Error('process.env.JWT_SECRET is undefined');

export const verifyToken =
    (req: NextRequest) =>
    {
        const cookie = res.cookies.get('session-token')!.value;
        const json = jwt.verify(token, jwtSecret);
	return JSON.parse(json);
    }
```

**Comment by GabenGar** [comment, marked as answer]
Validate in the [startup function](https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation#convention) instead.

**Reply by bbergeron-xlkey** [reply]
That works, but a word of notice for people who find this discussion; the startup function uses the Edge runtime, which blows big time.

**Reply by GabenGar** [reply]
Well the other way is to validate at call site, i.e. something like:
```ts
export const verifyToken =
    (req: NextRequest) =>
    {
        const jwtSecret = parseJWTSecret()
        const cookie = res.cookies.get('session-token')!.value;
        const json = jwt.verify(token, jwtSecret);
	return JSON.parse(json);
    }
    
function parseJWTSecret() {
  const jwtSecret = process.env.JWT_SECRET;
  if (jwtSecret === undefined) {
      throw new Error('process.env.JWT_SECRET is undefined');
  }
  
  return  jwtSecret;
}
```

**Reply by bbergeron-xlkey** [reply]
That's pretty much what I ended up doing although I could still check for missing env in the edge runtime (source: I hope so) but it's fine that way.
